### ISSUE TITLE
OpInfo Tracker

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#54261]
OpInfos are Python classes containing structured metadata, test directives, and sample inputs for PyTorch's operators. This data is used to automatically generate a variety of tests, including tests that operators work with a variety of systems, like autograd, forward autograd, torchscript, FX, and NNC correctly.

For more details on OpInfos and how to add them see the Github wiki article [here](https://github.com/pytorch/pytorch/wiki/Running-and-writing-tests#opinfos). If you have more questions ping @mruberry directly (like in a comment responding to this issue). 

**Priority Functions to OpInfo**
- [x] nn.functional.batch_norm (#63218)
- [x] nn.functional.layer_norm (#63276)
- [x] nn.functional.interpolate (https://github.com/pytorch/pytorch/pull/61956)
- [x] nn.functional.cross_entropy (https://github.com/pytorch/pytorch/pull/63547)
- [x] nn.functional.conv2d (https://github.com/pytorch/pytorch/pull/63517)
- [x] nn.functional.dropout (https://github.com/pytorch/pytorch/pull/62315)
- [x] nn.functional.linear (https://github.com/pytorch/pytorch/pull/61971)
- [x] nn.functional.max_pool2d (https://github.com/pytorch/pytorch/pull/63530)
- [x] nn.functional.nll_loss (https://github.com/pytorch/pytorch/pull/63854)
- [x] nn.functional.cosine_similarity (https://github.com/pytorch/pytorch/pull/62959)
- [x] nn.functional.embedding (https://github.com/pytorch/pytorch/pull/63959)
- [x] cat (see tests in test_autograd.py) (https://github.com/pytorch/pytorch/pull/62560)
- [x] block_diag (see tests in test_autograd.py) (https://github.com/pytorch/pytorch/pull/64993)
- [x] broadcast_tensors (see tests in test_autograd.py) (https://github.com/pytorch/pytorch/pull/64993)
- [ ] lobpcg (see tests in test_autograd.py)
- [ ] torch.linalg.tensor_solve (see tests in test_linalg.py)
- [x] igamma (see tests in test_autograd.py)
- [ ] as_strided (see tests in test_autograd.py)
- [ ] unbind (see tests in test_autograd.py)
- [ ] pdist (see tests in test_autograd.py)
- [x] [argsort](https://pytorch.org/docs/master/generated/torch.argsort.html?highlight=argsort#torch.argsort) (#65454)
- [x] [repeat_interleave](https://pytorch.org/docs/master/generated/torch.repeat_interleave.html?highlight=repeat_interleave#torch.repeat_interleave) (#65455)
- [x] [nn.functional.sigmoid](https://pytorch.org/docs/master/generated/torch.nn.functional.sigmoid.html?highlight=sigmoid#torch.nn.functional.sigmoid) (note that an OpInfo for torch.sigmoid exists)
- [x] nn.functional.tanh (note than an OpInfo for torch.tanh exists)
- [x] torch.nn.functional.conv1d (https://github.com/pytorch/pytorch/pull/67747)

**OpInfo Backlog**
- [ ] binary_cross_entropy_with_logits
- [ ] l1_loss (#65681)
- [ ] binary_cross_entropy
- [ ] elu
- [ ] upsample
- [ ] fold
- [ ] affine_grid
- [ ] max_pool1d (https://github.com/pytorch/pytorch/pull/67005)
- [ ] torch
- [ ] threshold
- [ ] smooth_l1_loss
- [ ] pairwise_distance (#65460)
- [ ] logsigmoid
- [ ] adaptive_max_pool2d
- [ ] pixel_shuffle (#65467)
- [ ] avg_pool3d https://github.com/pytorch/pytorch/pull/64214
- [ ] bilinear
- [ ] gumbel_softmax
- [ ] max_unpool2d (https://github.com/pytorch/pytorch/pull/67328)
- [ ] kl_div (#65469)
- [ ] ctc_loss
- [x] layer_norm (https://github.com/pytorch/pytorch/pull/63276)
- [ ] conv3d
- [ ] max_unpool3d (https://github.com/pytorch/pytorch/pull/67328)
- [ ] selu
- [ ] glu
- [ ] hardsigmoid
- [ ] upsample_bilinear
- [ ] max_pool3d (https://github.com/pytorch/pytorch/pull/67005)
- [ ] adaptive_avg_pool3d https://github.com/pytorch/pytorch/pull/64214
- [ ] instance_norm
- [ ] embedding_bag
- [ ] upsample_nearest
- [ ] avg_pool1d https://github.com/pytorch/pytorch/pull/64214
- [ ] prelu
- [ ] celu
- [ ] dropout2d (https://github.com/pytorch/pytorch/pull/67891)
- [ ] hinge_embedding_loss
- [ ] softsign
- [ ] max_unpool1d (https://github.com/pytorch/pytorch/pull/67328)
- [ ] silu
- [ ] softshrink
- [ ] leaky_relu_
- [ ] softmin
- [ ] channel_shuffle
- [ ] multilabel_margin_loss
- [ ] dropout3d
- [ ] multi_margin_loss
- [ ] lp_pool2d
- [ ] conv_transpose1d
- [ ] triplet_margin_loss
- [ ] tanhshrink
- [ ] adaptive_max_pool1d
- [ ] cosine_embedding_loss
- [ ] multi_head_attention_forward
- [ ] max_pool1d_with_indices (https://github.com/pytorch/pytorch/pull/67005)
- [ ] poisson_nll_loss
- [ ] margin_ranking_loss
- [ ] soft_margin_loss
- [ ] adaptive_max_pool3d
- [ ] group_norm
- [ ] local_response_norm
- [ ] multilabel_soft_margin_loss
- [ ] relu_
- [ ] alpha_dropout
- [ ] nn.functional.alpha_dropout (https://github.com/pytorch/pytorch/pull/67823)
- [ ] feature_alpha_dropout
- [ ] lp_pool1d
- [ ] adaptive_max_pool1d_with_indices
- [ ] adaptive_max_pool2d_with_indices
- [ ] adaptive_max_pool3d_with_indices
- [ ] fractional_max_pool2d
- [ ] fractional_max_pool2d_with_indices
- [ ] fractional_max_pool3d
- [ ] fractional_max_pool3d_with_indices
- [ ] max_pool2d_with_indices (https://github.com/pytorch/pytorch/pull/67005)
- [ ] max_pool3d_with_indices (https://github.com/pytorch/pytorch/pull/67005)
- [ ] handle_torch_function
- [ ] has_torch_function
- [ ] adaptive_avg_pool1d https://github.com/pytorch/pytorch/pull/64214
- [ ] pdist
- [ ] rrelu_
- [ ] elu_
- [ ] hardtanh_
- [ ] triplet_margin_with_distance_loss
- [ ] selu_
- [ ] pixel_unshuffle (#65468)
- [ ] conv_transpose3d
- [ ] gaussian_nll_loss
- [ ] celu_
- [ ] huber_loss
- [ ] mish
- [ ] threshold_
- [x] logical_and (https://github.com/pytorch/pytorch/pull/67178)
- [x] logical_or (https://github.com/pytorch/pytorch/pull/67178)
- [x] logical_xor (https://github.com/pytorch/pytorch/pull/67178)

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @dylanbespalko @mruberry @Lezcano @VitalyFedyunin @walterddr @mattip @kshitij12345

**Comment by imaginary-person** [comment]
In the example PR (https://github.com/pytorch/pytorch/pull/52579/files) referred to here, I had made some redundant assignments in `OpInfo`s. I added some new comments to that PR, so that a person new to the codebase or to the test architecture would not make a similar mistake while porting `OpInfo`s of other operators.

I'll also clean up `common_methods_invocations.py` a little via another PR by removing more redundant assignments in `OpInfo`s.

**Comment by Lezcano** [comment]
The operations prod, cumsum, cumprod and masked_scatter already have OpInfos.

**Comment by imaginary-person** [comment]
Please help resolve a low-priority query regarding OpInfo porting.

I was under the assumption that if `dtypesIfCPU` and `dtyesIfCUDA` are equal, then their assignment be skipped, and only the `dtypes` argument can be used while adding an OpInfo for an op in `common_methods_invocations.py`.

However, if I add `torch.float16` & `torch.bfloat16` to the `dtypes` list of `square` (after enabling `pow` on CPU for these dtypes), then the removal of either `dtypesIfCPU` or `dtypesIfCUDA` lists leads to several `test_unsupported_dtypes` tests failing.

So, should I not remove redundant `dtypesIfCPU` & `dtypesIfCUDA` assignments from `common_methods_invocations.py`?
Thank you!

**Comment by kshitij12345** [comment]
For UnaryUfuncInfo
https://github.com/pytorch/pytorch/blob/fd58ececab731803eceace70b985b35f9872b53c/torch/testing/_internal/common_methods_invocations.py#L310-L331

If you don't provide `dtypesIfCPU` or `dtypesIfCUDA`, it defaults to the dtypes in the constructor which may or may not be supported.

This is different from directly using OpInfo where if you don't provide `dtypesIfCPU` or `dtypesIfCUDA` then they are defaulted to `dtypes` (reference below),

https://github.com/pytorch/pytorch/blob/fd58ececab731803eceace70b985b35f9872b53c/torch/testing/_internal/common_methods_invocations.py#L135-L187

I think this causes confusion (often to me as well) and UnaryUfunc should have similar behavior as OpInfo.

@mruberry What do you think?

**Comment by imaginary-person** [comment]
Thanks, @kshitij12345, 
I should've paid attention that `square` has a `UnaryUfuncInfo`, and not an `OpInfo`.

**Comment by mruberry** [comment]
> I think this causes confusion (often to me as well) and UnaryUfunc should have similar behavior as OpInfo.
> 
> @mruberry What do you think?

Sounds reasonable. It was expedient to create the initial unary ufunc infos but I agree it's not helpful now.

**Comment by imaginary-person** [comment]
`addmm` already has an OpInfo but its `method_tests` had not been ported. Porting them via #55201.

**Comment by imaginary-person** [comment]
Hello @mruberry, all the non-docs CI checks had passed earlier today for #55201, which adds OpInfos for `add`, `mm`, and `addmm` (an additional `OpInfo` had to be added for `addmm`).

**Comment by imaginary-person** [comment]
I'll try to work on some of the remaining OpInfos.

**Comment by nkaretnikov** [comment]
i think these need to be marked as done:

```
 nn.functional.sigmoid (note that an OpInfo for torch.sigmoid exists)
 nn.functional.tanh (note than an OpInfo for torch.tanh exists)
```

both are defined as aliases in `UnaryUfuncInfo`:

```
https://github.com/pytorch/pytorch/blob/a31aea8eaa99a5ff72b5d002c206cd68d5467a5e/torch/testing/_internal/common_methods_invocations.py#L12598

https://github.com/pytorch/pytorch/blob/a31aea8eaa99a5ff72b5d002c206cd68d5467a5e/torch/testing/_internal/common_methods_invocations.py#L11146
```

(gh fails to show previews for large files sometimes)

```
    UnaryUfuncInfo('sigmoid',
                   aliases=('special.expit', 'nn.functional.sigmoid'),

[...]

    UnaryUfuncInfo('tanh',
                   ref=np.tanh,
                   aliases=('nn.functional.tanh',)
```

cc @mruberry

**Comment by nkaretnikov** [comment]
i'm trying to adding an opinfo for `torch.linalg.tensorsolve`. probably will take a bit since it's a first issue, so jfyi.

**Comment by Lezcano** [comment]
alas, `tensorsolve`'s already taken (https://github.com/pytorch/pytorch/pull/68810).
In general, all the `torch.linalg` operations are covered or about to be covered.

**Comment by nkaretnikov** [comment]
@Lezcano thx, i should have checked the issues (instead of simply relying on the list above). i'll pick something else.

**Comment by nkaretnikov** [comment]
i think i'll try adding `binary_cross_entropy`. it's missing and has no open issues/prs.

**Comment by nkaretnikov** [comment]
@mruberry could you add the following at the top of the first post (this is similar to what ed has [here](https://github.com/pytorch/pytorch/issues/55070))?

**WARNING: BEFORE STARTING ADDING AN OPINFO, CHECK OPEN PULL REQUESTS TO SEE IF SOMEONE IS WORKING ON IT ALREADY.**
https://github.com/pytorch/pytorch/pulls?q=is%3Apr+is%3Aopen+opinfo+in%3Atitle
(That is, search the page for the right operator instead of searching GitHub issues directly. Depending on how the title is written, the operator might not show up in results.)

**Comment by mruberry** [comment]
Actually I think we can close this particular issue at this point. It's pretty out of date and there are only a few remaining OpInfos that are being pursued using other tracking methods.
